tools/crm_rule.c

```diff
@@ -0,0 +1,266 @@
+
+    result = pe_eval_date_expression(match, effective_date);
+
+    if (result == date_after_range) {
+        printf("Rule %s is expired\n", rule_id);
+        rc = 1;
+    } else if (result == date_within_range) {
+        printf("Rule %s is still in effect\n", rule_id);
+        rc = 0;
+    } else if (result == date_before_range) {
+        printf("Rule %s has not yet taken effect\n", rule_id);
+        rc = 1;
+    } else {
+        printf("Could not determine whether rule %s is expired\n", rule_id);
+        rc = 1;
+    }
+
+bail:
+    free(xpath);
+    freeXpathObject(xpathObj);
+    pe_free_working_set(data_set);
+    return rc;
+}
+
+int
+main(int argc, char **argv)
+{
+    cib_t *cib_conn = NULL;
+
+    int flag = 0;
+    int option_index = 0;
+
+    char *rule_id = NULL;
+    crm_time_t *rule_date = NULL;
+
+    xmlNode *input = NULL;
+    char *input_xml = NULL;
+    char *input_file = getenv("CIB_file");
+
+    int rc = pcmk_ok;
+    crm_exit_t exit_code = CRM_EX_OK;
+
+    crm_log_cli_init("crm_rule");
+    crm_set_options(NULL, "[options]", long_options,
+                    "Tool for querying the state of rules");
+
+    while (flag >= 0) {
+        flag = crm_get_option(argc, argv, &option_index);
+        switch (flag) {
+            case -1:
+                break;
+
+            case 'V':
+                crm_bump_log_level(argc, argv);
+                break;
+
+            case '$':
+            case '?':
+                crm_help(flag, CRM_EX_OK);
+                break;
+
+            case 'c':
+                rule_mode = crm_rule_mode_check;
+                break;
+
+            case 'd':
+                rule_date = crm_time_new(optarg);
+                if (rule_date == NULL) {
+                    exit_code = CRM_EX_DATAERR;
+                    goto bail;
+                }
+
+                break;
+
+            case 'X':
+                input_xml = optarg;
+                break;
+
+            case 'r':
+                rule_id = strdup(optarg);
+                break;
+
+            default:
+                crm_help(flag, CRM_EX_OK);
+                break;
+        }
+    }
+
+    /* Check command line arguments before opening a connection to
+     * the CIB manager or doing anything else important.
+     */
+    if (rule_mode == crm_rule_mode_check) {
+        if (rule_id == NULL) {
+            CMD_ERR("--check requires use of --rule=\n");
+            crm_help(flag, CRM_EX_USAGE);
+        }
+    }
+
+    /* Set up some defaults. */
+    if (rule_date == NULL) {
+        rule_date = crm_time_new(NULL);
+    }
+
+    /* Where does the XML come from?  If one of various command line options were
+     * given, use those.  Otherwise, connect to the CIB and use that.
+     */
+    if (input_file != NULL) {
+        struct stat sb;
+
+        if (stat(input_file, &sb) == -1) {
+            CMD_ERR("cannot read input file %s: %s", input_file, strerror(errno));
+            exit_code = errno;
```

Comment:
This confuses the two. exit_code must be between 0 and 255, so errno shouldn't be used directly.  We created crm_errno2exit() to convert from an errno to a reasonable exit status, so that could be used here.

BTW this is why crm_exit_t was created -- in pacemaker 1.1, tools often exited with errno as a status. That has two main problems: first, errno values are not standardized, so exiting errno means someone can't write a portable script checking a particular exit status of a pacemaker tool; and second, errno can be outside the 0-255 range, so it winds up getting bitwise mapped into that range, creating weird, unpredictable exit statuses. crm_errno2exit() takes the errno and returns the closest semantically equivalent exit status, ensuring our exit status codes are reliable, even across OSes.